velox/exec/MergeJoin.cpp

```diff
@@ -814,11 +814,33 @@ RowVectorPtr MergeJoin::getOutput() {
     if (output != nullptr && output->size() > 0) {
       if (filter_) {
         output = applyFilter(output);
+
         if (output != nullptr) {
+          auto numRowsAfterFilter = output->size();
+
           for (const auto [channel, _] : filterInputToOutputChannel_) {
             filterInput_->childAt(channel).reset();
           }
-          return output;
+
+          if (finalOutputSize_ + numRowsAfterFilter <= outputBatchSize_) {
+            // Copy the output to the finalOutput
+            if (finalOutput_ == nullptr) {
+              finalOutput_ = std::move(output);
+            } else {
+              finalOutput_->resize(finalOutputSize_ + numRowsAfterFilter);
+              BaseVector::CopyRange range{
+                  0, finalOutputSize_, numRowsAfterFilter};
+              finalOutput_->copyRanges(output.get(), folly::Range(&range, 1));
```

Comment:
You can use `encodedVectorCopy` to preserve encoding: https://github.com/facebookincubator/velox/blob/ff41c351f7c9b75a64555bd5eedbb461c275c18d/velox/vector/EncodedVectorCopy.h#L58